plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/1Password/shell-plugins/sdk"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	confighelpers "github.com/99designs/aws-vault/v7/vault"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+type assumeRoleProvider struct {
+	confighelpers.AssumeRoleProvider
+	stsCacheWriter
+}
+
+func (p assumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.AssumeRoleProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
+	if in.Cache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, in.Cache)
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return &assumeRoleProvider{
+			AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+				StsClient:         getSTSClient(awsConfig.Region, NewMFASessionTokenProvider(awsConfig, in, out)),
+				RoleARN:           awsConfig.RoleARN,
+				RoleSessionName:   awsConfig.RoleSessionName,
+				ExternalID:        awsConfig.ExternalID,
+				Duration:          900 * time.Second, // minimum duration of 15 minutes
+				Tags:              awsConfig.SessionTags,
+				TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+				SourceIdentity:    awsConfig.SourceIdentity,
+				Mfa:               &confighelpers.Mfa{},
+			},
+			stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+		}
+	}
+
+	return &assumeRoleProvider{
+		AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+			StsClient:         getSTSClient(awsConfig.Region, NewMasterCredentialsProvider(in.ItemFields)),
+			RoleARN:           awsConfig.RoleARN,
+			RoleSessionName:   awsConfig.RoleSessionName,
+			ExternalID:        awsConfig.ExternalID,
+			Duration:          900 * time.Second, // minimum duration of 15 minutes
+			Tags:              awsConfig.SessionTags,
+			TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+			SourceIdentity:    awsConfig.SourceIdentity,
+			Mfa:               &confighelpers.Mfa{},
+		},
+		stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+	}
```

Comment:
There is a lot of duplication between these two returns which:
1. Makes it hard to see what's different
2. Makes it easier for unintended inconsistencies to be introduced in the future.

Reply:
Done, thanks!